src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1578,6 +1583,104 @@ private static bool HasAnonymousFunctionConversion(BoundExpression source, TypeS
 
             return IsAnonymousFunctionCompatibleWithType((UnboundLambda)source, destination) == LambdaConversionResult.Success;
         }
+
+        internal static ConstructibleCollectionTypeKind GetConstructibleCollectionType(CSharpCompilation compilation, TypeSymbol destination, out TypeSymbol? elementType)
+        {
+            Debug.Assert(compilation is { });
+
+            if (destination is ArrayTypeSymbol arrayType)
+            {
+                if (arrayType.IsSZArray)
+                {
+                    elementType = arrayType.ElementType;
+                    return ConstructibleCollectionTypeKind.Array;
+                }
+            }
+            else if (isSpanType(compilation, destination, WellKnownType.System_Span_T, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.Span;
+            }
+            else if (isSpanType(compilation, destination, WellKnownType.System_ReadOnlySpan_T, out elementType))
+            {
+                return ConstructibleCollectionTypeKind.ReadOnlySpan;
+            }
+            else if (implementsIEnumerable(compilation, destination))
```

Comment:
Do we assume the type also has an applicable Add method at this phase?

Reply:
No, we allow a target type without an `Add()` method for empty collection literals.